### PR TITLE
Update all dependencies (incl JS) and make use of thread_local

### PR DESCRIPTION
**Pull request by 9SMTM6**
Builds on the work of @th4s.

Also bumps rust editions, with a new edition coming in soon I think thats good to have more up to date. the projects also got updates to their JS dependencies.

wasm-pack is still out of date, stuck on #9.

I wasn't able to run the exact CI sequence on my laptop since I dont run ubuntu, and playwright install doesnt seem to like that, but with a bit of nudging it seemed to work.

closes #10

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -10,7 +10,7 @@ jobs:
     steps:
     - uses: actions/checkout@v3
     - name: Setup nightly Rust
-      run: rustup toolchain install nightly-2022-12-12 --profile minimal --target wasm32-unknown-unknown --component rust-src
+      run: rustup toolchain install nightly-2024-08-02 --profile minimal --target wasm32-unknown-unknown --component rust-src
```
Comment (RReverser) [inline .github/workflows/ci.yml:13]: These shouldn't be necessary with `rust-toolchain.toml` now, right?

File: test/rust-toolchain.toml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# If you see this, run "rustup self update" to get rustup 1.23 or newer.
+
+# NOTE: above comment is for older `rustup` (before TOML support was added),
+# which will treat the first line as the toolchain name, and therefore show it
+# to the user in the error, instead of "error: invalid channel name '[toolchain]'".
+
+[toolchain]
+# channel = "stable"
+# If you change this, adjust the toolchain in the pipeline accordingly, otherwise the pipelines will and up very slow.
+channel = "nightly-2024-08-02"
```
Comment (RReverser) [inline test/rust-toolchain.toml:10]: Let's also set `profile = "minimal"` as it was specified in CI config before.

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -14,7 +14,7 @@ jobs:
     - name: Setup stable Rust
       run: rustup target add wasm32-unknown-unknown
     - name: Setup nightly Rust
-      run: rustup toolchain install nightly-2022-12-12 --profile minimal --target wasm32-unknown-unknown --component rust-src
+      run: rustup toolchain install nightly-2024-08-02 --profile minimal --target wasm32-unknown-unknown --component rust-src
```
Comment (RReverser) [inline .github/workflows/deploy.yml:17]: Same here, let's just let `rust-toolchain.toml` take care of these for us now.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (9SMTM6) [reply .github/workflows/ci.yml:13]: probably right, I just blindly replaced the toolchain here.

Gonna test it.

**Comment by 9SMTM6** [comment]
@RReverser not sure why that test failed. It ran fine on my fork.

I think it was a bit flaky for me when I ran it locally, before some other changes, I thought these fixed it. Were they ever flaky in the past?

**Comment by 9SMTM6** [comment]
@RReverser Ive now fixed the more recent issues in the CI, not super elegant, but IDK another way to trigger the installation according to toolchain.toml.

The issues we saw in the first CI run in this repo (2 days ago, commit d1816e5bc2bf928ff5442355c04500a381d66a41) have not really been addressed specifically, so they might reappear.

**Comment by 9SMTM6** [comment]
I've been also so free to remove the CLA reference I mentioned in #10.

**Comment by RReverser** [comment]
Weird. I don't think it was flaky in the past, no, but entirely plausible something on their server has changed.

Thanks!

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -9,11 +9,14 @@ jobs:
...
-    - name: Setup nightly Rust
-      run: rustup toolchain install nightly-2022-12-12 --profile minimal --target wasm32-unknown-unknown --component rust-src
     - name: Install npm deps
       working-directory: ./test
       run: npm ci
+    # For some reason isntalling the rust toolchain via wasm-pack leads to CI failures.
+    # This check will read the top level rustup-toolchain.toml and install the required stuff.
+    - name: Trigger toolchain installation
+      working-directory: ./test
+      run: cargo check
```
Comment (RReverser) [inline .github/workflows/ci.yml:19]: Would `cargo fetch` be sufficient here? That way we would still have clear separation between "download dependencies" and "actually build" steps.

Comment (9SMTM6) [reply .github/workflows/ci.yml]: Huh, i wasnt aware of that command. I could not find something with similar terms to install (npm) etc, and when I found out about cargo-chef I took that as a replacement for what cargo fetch seems to do.

But yeah, this seems to be sufficient according to the test commit I just did!

**Comment by RReverser** [comment]
Awesome, thanks @9SMTM6 and @th4s both!

**Comment by RReverser** [comment]
> warning: unused return value of `clone` that must be used
  --> src/lib.rs:58:5
   |
58 |     wasm_bindgen::link_to!(module = "/src/workerHelpers.worker.js");
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |

Can you suppress this please? I imagine all users will get this warning when building this as a transitive dependency, and it's just noise at this point.

**Pull request merged**